src/coreclr/jit/optimizer.cpp

```diff
@@ -3043,11 +2945,119 @@ bool Compiler::optCreatePreheader(FlowGraphNaturalLoop* loop, bool* splitHeader)
         fgReplaceJumpTarget(enterBlock, header, preheader);
     }
 
+    loop->SetEntryEdge(newEdge);
+
     optSetWeightForPreheaderOrExit(loop, preheader);
 
     return true;
 }
 
+//-----------------------------------------------------------------------------
+// optSplitHeaderIfNecessary: If preheader and header are in different try
+//  regions, split the header to put it into the same try region as the preheader
+//
+// Parameters:
+//   loop - loop that may need header splitting
+//
+// Returns:
+//   True if the header was split
+//
+// Notes:
+//   Ensures that no loop header is also a try entry.
+//
+bool Compiler::optSplitHeaderIfNecessary(FlowGraphNaturalLoop* loop)
+{
+    BasicBlock* header    = loop->GetHeader();
+    BasicBlock* preheader = loop->GetPreheader();
+
+    if (BasicBlock::sameTryRegion(header, preheader))
+    {
+        assert(!bbIsTryBeg(header));
+        return false;
+    }
+
+    // If the preheader and header are in different try regions,
+    // the header should be a try entry.
+    //
+    assert(bbIsTryBeg(header));
+
+    JITDUMP("Splitting " FMT_LP " header / try entry " FMT_BB "\n", loop->GetIndex(), header->bbNum);
+
+    Statement* const firstStmt   = header->firstStmt();
+    BasicBlock*      newTryEntry = nullptr;
+
+    if (firstStmt == nullptr)
+    {
+        // Empty header
+        //
+        newTryEntry = fgSplitBlockAtEnd(header);
+    }
+    else
+    {
+        // Non-empty header.
+        //
+        Statement* const lastStmt      = header->lastStmt();
+        bool const       hasTerminator = header->HasTerminator();
+        Statement* const stopStmt      = hasTerminator ? lastStmt : nullptr;
+        Statement*       splitBefore   = firstStmt;
+
+        while ((splitBefore != stopStmt) && (splitBefore->GetRootNode()->gtFlags & GTF_EXCEPT) == 0)
+        {
+            splitBefore = splitBefore->GetNextStmt();
+        }
```

Comment:
Does this need to check for `GTF_CALL` too? I think your recent try-removal passes did that to be safe.

Reply:
Since we're post-morph I would hope `GTF_EXCEPT` is sufficient -- those other phases run very early where we're not yet doing flags validation.

Reply:
I'm mostly worried that certain classes of exceptions that we sometimes allow reordering (like `OutOfMemoryException`) would suddenly end up being moved out of a handler. There may be some cases where we intentionally set only `GTF_CALL` but not `GTF_EXCEPT`, but I'm not sure.

Reply:
It does look like allocation helpers are considered "no throw" and only get `GTF_CALL`, so we probably do need to check for that here too. Seems like we are moving to a state where we need to start differentiating between "throws reorderable exceptions" and "truly does not throw", since only the latter can be moved out of a try region.

Reply:
Ok, added GTF_CALL.